en/concept/events_interface.md

```diff
@@ -0,0 +1,133 @@
+
+The above is a minimal example, this is a more extensive one:
+```cpp
+uint8_t arg1 = 0;
+float arg2 = -1.f;
+/* EVENT
+ * @description
+ * This is the detailed event description.
+ *
+ * - value of arg1: {1}
+ * - value of arg2: {2:.1}
+ *
+ * <profile name="dev">
+ * (This paragraph is only meant to be shown to developers).
+ * This behavior can be configured with the parameter <param>COM_EXAMPLE</param>.
+ * </profile>
+ *
+ * Link to documentation: <a>https://docs.px4.io</a>
+ */
+events::send<uint8_t, float>(events::ID("event_name"),
+	{events::Log::Error, events::LogInternal::Info}, "Event Message", arg1, arg2);
+```
+
+Explanations and requirements:
+- `/* EVENT`: this is an indicator for the parser that there is additional metadata following as comment.
+- **event_name**: the event name.
+  - must be unique within the whole source code of PX4.
+    As a general convention, prefix it with the module name, or the source file for larger modules.
+  - must be a valid variable name, i.e. must not contain spaces, colons, etc.
+  - from that name, a 24 bit event ID is derived using a hash function.
+    This means as long as the event name stays the same, so will the ID.
+- **Log Level**:
+  - valid log levels are (in descending importance):
+    ```
+	Emergency,
+	Alert,
+	Critical,
+	Error,
+	Warning,
+	Notice,
+	Info,
+	Debug,
+	Disabled,
+	```
+  - Above we specify a separate external and internal log level, which are the levels displayed to GCS users and in the log file, respectively: `{events::Log::Error, events::LogInternal::Info}`.
+    For the majority of cases you can pass a single log level, and this will be used for both exernal and internal cases.
+	There are cases it makes sense to have two different log levels.
+	For example an RTL failsafe action: the user should see it as Warning/Error, whereas in the log, it is an expected system response, so it can be set to `Info`.
+- **Event Message**:
+  - Single-line, short description of the event.
+    It may contain arguments (see below).
+- **Event Description**:
+  - detailed, optional event description.
+  - can be multiple lines/paragraphs.
+
+#### Arguments and Enums
+
+Events can have a fixed set of arguments that can be shown in the message or description.
+
+Valid types: `uint8_t`, `int8_t`, `uint16_t`, `int16_t`, `uint32_t`, `int32_t`, `uint64_t`, `int64_t` and `float`.
+
+It is also possible to define and use custom enumerations.
+These are defined in [src/lib/events/enums.json](https://github.com/PX4/PX4-Autopilot/blob/master/src/lib/events/enums.json), and can then be used as event argument in the form of `events::send<events::px4::enums::my_enum_t>(...)`.
+
+#### Text format
+
+Text format for event message description:
+- characters can be escaped with \\
+
+  These have to be escaped: '\\\\', '\\<', '\\{'.
+- supported tags:
+  - Profiles: `<profile name="[!]NAME">CONTENT</profile>`
+
+    `CONTENT` will only be shown if the name matches the configured profile.
```

Comment:
What is a profile? When will this information be shown? 

My guess is that the information in the event description gets somehow parsed "somewhere" and that you can configure various parts that will go to various audiences. I assume this is different parts of docs? Or is is GCS? 

I guess what I'm saying is that I THINK this whole section is about the documentation part, but I'm not sure. I think perhaps a real example here would be useful.

Reply:
The concept needs a bit of explanation (on the events interface level, but could be repeated here): It's used to show certan text to specific audiences (on the GCS). For example you would only show 'this check can be disabled via parameter xy' to a developer, but not an end-user. The GCS then has a config option if the user is a developer or end-user.
The is an example already above.

Reply:
> The is an example already above.

The problem is that it feels ambiguous. For example `* - value of arg1: {1}` could mean lots of things - i.e. perhaps you mean literal text `value of arg1:`, or perhaps it is placeholder text meaning `arg_1_is_a_foo`. A real example would make it clear which of them. 

With respect to profiles - fair enough. I guess we can link/revisit once those docs exist.

Reply:
I don't follow your ambiguity, but feel free to add a real example (or change as you see fit).

Reply:
Catch 22 - how can I add a real example if I don't understand the syntax. I searched and there aren't any in the codebase where there are descriptions for arguments.
The only example is one that has just the single line description.
```
/* EVENT
 * @description
 * Land immediately and check the system.
 */
events::send<events::px4::enums::sensor_type_t, uint8_t, events::px4::enums::sensor_failover_reason_t>(
  events::ID("sensor_failover"), events::Log::Emergency, 
  "{1} sensor #{2} failure: {3}", sensor_type, failover_index,
						failover_reason);
```

So for the above, what would the arguments look like - e.g. what should be in place in this event for these in your pseudo example?:
```
 * - value of arg1: {1}
 * - value of arg2: {2:.1}
 ```

Reply:
Sorry I still don't follow.
In the example, the user would see this:
```
 * - value of arg1: 0
 * - value of arg2: -1.0
```

Reply:
Thanks!

Reply:
So you think we should have a different example?

Reply:
@bkueng No, I completely don't get this, but I'm assuming I'm just being thick, and it will be obvious. Perhaps @julianoes could read it and see if does?

What I had hoped for was a real event in our codebase that had argument documentation - basically used the entire syntax in a real case. With that the docs could link to the real example so even if I don't get it, others might. 
I think perhaps this will be more clear once the mavlink docs are present - my confusion is probably indicative that I don't fully understand/remember what we're trying to do.

For example, I'm see from the [proposal](https://docs.google.com/document/d/18qdDgfML97lItom09MJhngYnFzAm1zFdmlCKG7TaBpg/edit#heading=h.l8v2vka1vs88) and the [EVENT](https://mavlink.io/en/messages/common.html#EVENT) message that the event has a 40 byte buffer for the arguments. My original assumption was that this means 40 arguments, but now I think maybe the buffer is "packed" with the arguments, and it is up to the event definitions to define how this is done.

IF that is correct, then shouldn't the event code be dynamically generated from the event definition file rather than explicitly written in our code? Then if we want to send an event or receive and event we use handy methods autogenerated for us - rather than having to hand code everything like this? I.e. the same way we unpack messages?

ANyway, still confused. Sorry.

Reply:
Let me try again using the real example ....
```
/* EVENT
 * @description
 * Land immediately and check the system.
 */
events::send<events::px4::enums::sensor_type_t, uint8_t, events::px4::enums::sensor_failover_reason_t>(
  events::ID("sensor_failover"), events::Log::Emergency, 
  "{1} sensor #{2} failure: {3}", sensor_type, failover_index,
						failover_reason);
```

Here
- we just have the one line description with the text "Land immediately and check the system.". 
  - The GCS presumably has a shared understanding based on the event id (`sensor_failover`) and can associate the above message from component information metadata with the event.
- We have three arguments, sensor_type, failover_index, failover_reason, we know they are "bytes" from the `send` signature. 
- We have the message itself  `{1} sensor #{2} failure: {3}` (so `{sensor_type1} sensor #{failover_index} failure: {failover_reason}`). I assume this is the format of the message in the metadata which is logged - i.e. this goes into metadata to allow reconstruction of the message, but is not actually sent.
- So there is no argument information in this definition.  Looking at your original example my guess is that this provides information about how the argument is supposed to be rendered at the other end and in text - it isn't actually about a _particular_ value. So this tells us that arg 1 is just rendered as whatever value is passed into the argument. It tells us that the second argument is presented as a number with `NUM_DECIMAL_DIGITS` (i.e. it is a number like 2.1, not 2.11). Is that right? So the essentially in this real example above we're just saying that this is the default - there is no formatting - it is just a number without decimal points or units?
  ```
  * - value of arg1: {1}
  * - value of arg2: {2:.1}
 ```

Reply:
> now I think maybe the buffer is "packed" with the arguments, and it is up to the event definitions to define how this is done.

Yes that is correct.

> This tells us that arg 1 is just rendered as whatever value is passed into the argument. It tells us that the second argument is presented as a number with NUM_DECIMAL_DIGITS (i.e. it is a number like 2.1, not 2.11). Is that right?

Correct

> so if the value is rendered without decimal points or units then you just omit the arguments.

No, the decimal points is just to configure the rendering. For int's you would not need it.

In general there's 2 use-cases:
- event definition in px4 source: in that case you don't typically care about the arguments, but just about the message (which may or may not print the argument(s)). In that case the parser will just fill in the args automatically, no special code generation needed.
- for events in common namespace, there is code generation, and you have the equivalent as in mavlink to get the arguments (not documented here, as not yet used). But you can still just use it as in the previous case when only caring about the message.

Does that explain it a bit better?

Reply:
> Does that explain it a bit better?

Definitely getting there! This is part of the story that I was missing - there is are two ways to define events. I had assumed that we would follow a MAVLINK dialect-like model for defining dialect/non-standard events. In other words you'd define the events in JSON or XML or whatever for your dialect and that would generate the methods to both send and extract your event and also the metadata.

Instead, my understanding is now that the common **events** are predefined and some PX4 **enums** are predefined. However for PX4 (say) events are generated into the required metadata at build time. If the value is a PX4 enum then the enum itself defines the size of the data for the argument, if not then this can presumably be extracted from the method signature. Either way, this is turned into event metadata a bit like https://github.com/mavlink/libevents/blob/master/events/common.json#L291 for the PX4 event (presumably if this uses enums in common or PX4 they are cross linked magically). Right?

Why not just define the whole event in the PX4 .json file? I can see it might make things faster to define, but presumably QGC would have to write custom code to extract the PX4 events - which seems a little harder?

> > so if the value is rendered without decimal points or units then you just omit the arguments.

> No, the decimal points is just to configure the rendering. For int's you would not need it.

Not sure how this clarification differs from what I said.
- The `* - value of arg2:` defines how the arg2 will be rendered. So imagine this is  `* - value of arg2:` {2:.2m}`, that says whenever a message has the placeholder `{2}` it will render the argument in the format indicated. So assuming arg 2 is a real number, 5.4556 and we had a message definition "The distance is: {2}"it will be rendered as "The distance is: 5.45m`
- You only need decimal for reals. YOu might need the units for any of them. 
- Right?

I think perhaps "value of arg" was a bit unhelpful for the identifier - it wasn't even clear to me at first this was actually an identifier from the point of view of the parser. Might have been more intuitive as "@arg1_format_string:`
---

Is there any more context for params and URLs? Sometimes we use the term param meaning arguments? What are URLS and params for?/how are they used?/how do they map to the generated event information? 

Above tells us they exist, but not sure why they are needed/when they are needed, how they map to a common event (say).

Reply:
> Right?

Correct.

> Why not just define the whole event in the PX4 .json file?

Because of maintainability & compatibility: I've seen it myself already, if I had to edit an external file for 100+ events, to then use them in code, it's just too cumbersome. And also harder to trace an error message in code.
Then we want to avoid QGC knowing about as many events as possible, so we can freely update events on the PX4 side.

> You only need decimal for reals. YOu might need the units for any of them.

Yes

>  "value of arg" was a bit unhelpful for the identifier

It's not the identifier. `{x}` is, where x is the argument index.

Reply:
> Is there any more context for params and URLs? Sometimes we use the term param meaning arguments? What are URLS and params for?/how are they used?/how do they map to the generated event information?

The idea is that you can have urls and px4 params in descriptions. And there's a special syntax so that the parser can generate a link out of it for displaying (this is application-specific). E.g. clicking on a param will open the param editor.

Reply:
>> "value of arg" was a bit unhelpful for the identifier

> It's not the identifier. {x} is, where x is the argument index.

In a real example, might the actual text that appears be `* - value of arg2: {2:.1}`? What I mean by saying "identifier" is that the text `value of arg` is something the parser looks for in order to tell that the format of the string used for presenting arguments is what follows. If that is incorrect then I reiterate, need a pointer to code that shows an event that includes use of a URL. 

Can you add an actual case of URL and param in your example that we can refer to?

Reply:
> Because of maintainability & compatibility: ...

Makes a lot of sense. Does/can QGC still automatically ease extraction of the actual arguments and strings from the metadata or do you still have to create code to do that in QGC every time you add/update a new PX4 side event?

Reply:
> What I mean by saying "identifier" is that the text value of arg is something the parser looks for in order to tell that the format of the string used for presenting arguments is what follows.

Yes that is incorrect. The parser only cares about `{...}` and tags `<...>...</...>`.
The url example is also in there above: `<a>https://docs.px4.io</a>`, which would just turn into plain text `https://docs.px4.io` by the default parser config.

> Does/can QGC still automatically ease extraction of the actual arguments and strings from the metadata or do you still have to create code to do that in QGC every time you add/update a new PX4 side event?

It's all automatic. Meaning you can use any version of QGC that supports the events interface.

Reply:
Finally I get it. The description is whatever text you like with some placeholders for urls that will get generated in by the parser. Some tags are also interpreted in a particular way - e.g. `<a>`